plugins/incremental-ingestion-backend/src/database/IncrementalIngestionDatabaseManager.ts

```diff
@@ -0,0 +1,592 @@
+              .delete()
+              .whereIn(
+                'ingestion_id',
+                ingestionIDs.map(entry => entry.id),
+              )
+          : 0;
+
+      const ingestionsDeleted = await tx('ingestion.ingestions')
+        .delete()
+        .where('provider_name', provider);
+
+      const next_action_at = new Date();
+      next_action_at.setTime(next_action_at.getTime() + 24 * 60 * 60 * 1000);
+
+      await this.insertIngestionRecord({
+        id: v4(),
+        next_action: 'rest',
+        provider_name: provider,
+        next_action_at,
+        ingestion_completed_at: new Date(),
+        status: 'resting',
+        completion_ticket: v4(),
+      });
+
+      return { provider, ingestionsDeleted, marksDeleted, markEntitiesDeleted };
+    });
+  }
+
+  /**
+   * Creates a new ingestion record.
+   * @param provider - string
+   * @returns A new ingestion record
+   */
+  async createProviderIngestionRecord(provider: string) {
+    const ingestionId = v4();
+    const nextAction = 'ingest';
+    try {
+      await this.insertIngestionRecord({
+        id: ingestionId,
+        next_action: nextAction,
+        provider_name: provider,
+        status: 'bursting',
+        completion_ticket: 'open',
+      });
+      return { ingestionId, nextAction, attempts: 0, nextActionAt: Date.now() };
+    } catch (_e) {
+      // Creating the ingestion record failed. Return undefined.
+      return undefined;
+    }
+  }
+
+  /**
+   * Computes which entities to remove, if any, at the end of a burst.
+   * @param provider - string
+   * @param ingestionId - string
+   * @returns All entities to remove for this burst.
+   */
+  async computeRemoved(provider: string, ingestionId: string) {
+    return await this.client.transaction(async tx => {
+      const removed: { entity: string; ref: string }[] = await tx(
+        'final_entities',
+      )
+        .select(
+          tx.ref('final_entity').as('entity'),
+          tx.ref('refresh_state.entity_ref').as('ref'),
+        )
+        .join(
+          'refresh_state',
+          'refresh_state.entity_id',
+          'final_entities.entity_id',
+        )
+        .whereRaw(
+          `((final_entity::json #>> '{metadata, annotations, ${INCREMENTAL_ENTITY_PROVIDER_ANNOTATION}}')) = ?`,
+          [provider],
+        )
+        .whereNotIn(
+          'entity_ref',
+          tx('ingestion.ingestion_marks')
+            .join(
+              'ingestion.ingestion_mark_entities',
+              'ingestion.ingestion_marks.id',
+              'ingestion.ingestion_mark_entities.ingestion_mark_id',
+            )
+            .select('ingestion.ingestion_mark_entities.ref')
+            .where('ingestion.ingestion_marks.ingestion_id', ingestionId),
+        );
+      return removed.map(entity => {
+        return { entity: JSON.parse(entity.entity) };
+      });
+    });
+  }
+
+  /**
+   * Performs a lookup of all providers that have duplicate active ingestion records.
```

Comment:
Should there be a UNIQUE constraint on the table instead?

Reply:
This method is left over from before we added the unique constraint that is present in the current migration. I'll have to think about whether this is still a concern.